# Numerical Integration

## 1. Trapezoidal Rule

### Formula

 For a partitition $x_0 = a < x_1 < \cdots < x_n = b$, the trapezoid rule formula simplifies to  
 
 $$\int_a^b\,f(x)\,dx \approx \frac{1}{2} \sum_{i=1}^n (f(x_i) + f(x_{i-1})) (x_i - x_{i-1})$$

 In particular, given an integer $N$, the trapezoid rule for $N$ sub-intervals of $[a,\,b]$ of equal length is  
 
 $$T_N(f) = \frac{\Delta x}{2} \sum_{i = 1}^N (f(x_i) + f(x_{i-1}))$$
 where $\Delta x = \frac{b - a}{N}$ is the length of the sub-intervals and $x_i = a + i\,\Delta x$.

 Notice that the trapezoid is the average of the left and right Riemann sums  
 
 $$\begin{aligned} T_N(f) &= \frac{\Delta x}{2} \sum_{i = 1}^N (f(x_i) + f(x_{i-1}))\\&= \frac{1}{2} \left( \sum_{i=1}^N\,f(x_i)\,\Delta x + \sum_{i=1}^N\,f(x_{i-1})\,\Delta x \right)\end{aligned}$$

### Implementation

In [2]:
import numpy as np
import matplotlib.pyplot as plt

In [3]:
def trapzoidal(
    f,
    a,
    b,
    N=50
):
    
    '''
    Parameters
    ----------
    f : function
        Vectorized function of a single variable
    a , b : numbers
        Interval of integration [a,b]
    N : integer
        Number of sub-intervals of [a,b]

    Returns
    -------
    T : float
        Approximation of the integral of f(x) from a to b using the
        trapezoid rule with N subintervals of equal length.
    '''
    
    '''
    Hint:
        1. Compute the equivalent points
        2. Compute the value of f(x)
        3. Prepare right/left endpoints
        4. Compute the length of each sub-interval
        5. Compute the integal
    '''

    x = np.linspace(a,b,N+1) 
    y = f(x)
    y_right = y[1:] 
    y_left = y[:-1]
    dx = (b-a)/N 
    T = dx/2 * np.sum(y_left + y_right) 

    return T

### Example

Compute the integral
 $$\int_0^{\pi/2}\,\sin x\,dx$$

 Note that we already know the real answer is $1$.

In [4]:
def sin(x):
    return np.sin(x)
approximation = trapzoidal(np.sin,0,np.pi/2,1000)
print('The approximate value by trapzoidal rule is', approximation)
print('The error is', abs(1 - approximation))

The approximate value by trapzoidal rule is 0.9999997943832332
The error is 2.056167668351705e-07


## 2. Simpson's Rule

### Formula
Simpson's Rule uses a quadratic polynomial on each subinterval of a partition to approximate the function $f(x)$ and to compute the definite integral.

This is an improvement over the trapezoid rule which approximates $f(x)$ by a straight line on each subinterval of a partition.

The formula for Simpson's rule is
 $$S_N(f) = \frac{\Delta x}{3} \sum_{i = 1}^{N/2} (f(x_{2i - 2}) + 4f(x_{2i - 1}) + f(x_{2i}))$$

where $N$ is an even number of sub-intervals of $[a,\,b],\,\Delta x = \frac{b - a}{N}$ and $x_i = a + i \Delta x$.

### Implementation

In [5]:
def simpson(
    f,
    a,
    b,
    N=50
):
    
    '''
    Parameters
    ----------
    f : function
        Vectorized function of a single variable
    a , b : numbers
        Interval of integration [a,b]
    N : (even) integer
        Number of subintervals of [a,b]

    Returns
    -------
    S : float
        Approximation of the integral of f(x) from a to b using
        Simpson's rule with N subintervals of equal length.
    '''
    
    '''
    Hint:
        1. Check whether 'N' is even
        2. Compute the length of each sub-interval
        3. Compute the value of f(x)
        4. Compute the integal
    '''
    if N%2 ==1:
        raise ValueError("N must be an integer.")
        
    dx = (b-a)/N
    x = np.linspace(a,b, N+1) 
    y=f(x)
    S = dx/3 * np.sum(y[0:-1:2] + 4*y[1::2] + y[2::2]) 
    
    return S

### Example

Compute the integral
 $$\int_0^{\pi/2}\,\sin x\,dx$$

 Note that we already know the real answer is $1$.

In [6]:
approximation = simpson(np.sin, 0, np.pi/2, 100)
print('The approximate value is', approximation)
print('The error is', abs(1 - approximation))

The approximate value is 1.000000000338236
The error is 3.382361057902017e-10


## 3. Gaussian–Legendre Quadrature

### Implementation

In [7]:
from scipy.special.orthogonal import p_roots

In [8]:
def gauss(  
    f,
    n,
    a,
    b
):
    '''
    Parameters
    ----------
    f : function
        Vectorized function of a single variable
    n : integer
        Number of points
    a , b : numbers
        Interval of integration [a,b]

    Returns
    -------
    G : float
        Approximation of the integral of f(x) from a to b using the
        Gaussian–Legendre quadrature rule with N points.
    '''
    
    [x,w]=p_roots(n)
    G = 0.5*(b-a)*sum(w*f(0.5*(b-a)*x+0.5*(b+a)))
   
    return G

### Example

Compute the integral
 $$\int_0^{\pi/2}\,\sin x\,dx$$

 Note that we already know the real answer is $1$.

In [9]:
for n in range(1, 11):
    approximation = gauss(np.sin, n, 0, np.pi/2)
    print('The approximate value of n =', n, 'is', approximation)

The approximate value of n = 1 is 1.1107207345395915
The approximate value of n = 2 is 0.9984726134041148
The approximate value of n = 3 is 1.0000081215554981
The approximate value of n = 4 is 0.999999977197115
The approximate value of n = 5 is 1.0000000000395652
The approximate value of n = 6 is 0.9999999999999535
The approximate value of n = 7 is 0.9999999999999999
The approximate value of n = 8 is 1.0
The approximate value of n = 9 is 1.0
The approximate value of n = 10 is 1.0000000000000002


## 4. Other Quadrature Formulae

### Example
 Let $I(f)$ be a define integral defined by
 $$I(f) = \int_0^1 f(x) dx,$$
 and consider the quadrature formula
 $$\hat{I}(f) = \alpha_1 f(0) + \alpha_2 f(1) + \alpha_3 f'(0) \quad \quad \quad (*)$$
 for approximation of $I(f)$.

### Part 1. 

#### We want to determine the coefficients $\alpha_j$ for $j = 1,\,2,\,3$ in such a way that $\hat{I}$ has the degree of exactness $r = 2$. Here the degree of exactness $r$ is to find $r$ such that
#### $$I(x^k)=\hat{I}(x^j) \quad \text{for} \quad k = 0,\,1,\dots,\,r \quad \text{and} \quad \hat{I}(x^j) \neq I(x^j) \quad \text{for} \quad j > r, $$
#### where $x^j$ denote the $j$-th power of $x$.  
 
 ---
 
 Since $\hat{I}$ has the degree of exactness $r = 2$
$$I(f) = \int_0^1 f(x) dx = \hat{I}(f) = \alpha_1 f(0) + \alpha_2 f(1) + \alpha_3 f'(0) $$
for $f(x)=1$, $f(x)=x$ and $f(x)=x^2, $  
and the corresponding differential of $f(x)$ is $f'(x) = 0, 1, 2x$ respectively.

For $f(x)=1$,  $$\int_0^1 1 dx = 1 = \alpha_1 \cdot 1 + \alpha_2 \cdot 1 + \alpha_3 \cdot 0 $$
For $f(x)=x$,  $$\int_0^1 x dx = {1\over2} = \alpha_1 \cdot 0 + \alpha_2 \cdot 1 + \alpha_3 \cdot 1$$  
For $f(x)=x^2$,  $$\int_0^1 x^2 dx = {1\over3} = \alpha_1 \cdot 0 + \alpha_2 \cdot 1 + \alpha_3 \cdot 0$$

Solving the set of equations, and we can get $$\alpha_1={2\over3},\alpha_2={1\over3}, \alpha_3={1\over6}$$

### Part 2. 

#### We want to find an apppropriate expression for the error $E(f) = I(f) - \hat{I}(f)$.

---
Since $$I(f) = \int_0^1 f(x) dx, \,\,\,\,\, \hat{I}(f) = {2\over3} f(0) + {1\over3} f(1) + {1\over6} f'(0) $$
and $$\exists\,\, n=2\,\,  s.t. \,\, \hat{I}(P_{i})=I(P_{i}) \,\,\,\, \forall \,i\leq 2$$ 
By Peano Kernel Theorem, $$E(f)=\int_0^1 {1\over2!}f^{(3)}(t)\,K(t)\, dt,\,\,\,\,where\,\,\, K(t)=E((x-t)^2_{+})$$
Since $$(x-t)^n_{+}=\begin{cases} 
(x-t)^n\,\,if \,\,t\leq x \\ 
0\;\;\;\;\;\;\;\;\;\;if\,\, t> x\\ 
\end{cases}
\implies {d\over dx} (x-t)^n_{+}=\begin{cases} 
n(x-t)^{n-1}\,\,if \,\,t\leq x \\ 
0\;\;\;\;\;\;\;\;\;\;\;\;\;\;if\,\, t> x\\ 
\end{cases}$$
If we let $$f(x)=(x-t)^n_{+}$$
then $$f(0)=0,\;f(1)=(1-t)^2,\;f'(0)=0\;\;\;\;since\;\;\; t\in [0,1]$$
So we can get $$\begin{aligned}
K(t) &=E((x-t)^2_{+}) \; = I((x-t)^2_{+}) - \hat{I}((x-t)^2_{+})\\
&= \int_0^1 (x-t)^2_{+}\, dx\,-({2\over 3}\times 0 + {1\over 3}(1-t)^2 +{1\over6}\times 0)\\
&= \int_t^1 (x-t)^2\, dx\,-{1\over 3}(1-t)^2\\
&= {1\over 3}(x-t)^3\bigm|_{x=t}^1\,-{1\over 3}(1-t)^2\\
&= -{1\over 3}t(1-t)^2\\
\end{aligned}$$  

Since $K(t)<0\;\;\;\forall t\in [0,1]$  (K(t) does not change sign), by MVT  

$$E(f)=\int_0^1 {1\over2} f^{(3)}(t)\,K(t)\,dt\;={1\over2} f^{(3)}(\xi)\int_0^1\,K(t)\,dt\;\;\;for\;some\;\;\xi \in [0,1]$$
Plus, $$\int_0^1\,K(t)\,dt = \int_0^1\, -{1\over 3}t(1-t)^2 \,dt = -{1\over 3}\int_0^1\, t-2t^2+t^3\;dt=-{1\over 3}({1\over 2}t^2-{2\over 3}t^3+{1\over 4}t^4)\bigm|_0^1 = -{1\over 36}$$
Therefore, $$E(f)={1\over 2}f^{(3)}(\xi)(-{1\over 36})=-{1\over 72}f^{(3)}(\xi)\;\;\;for\;some\;\;\xi\in[0,1]$$

---
### Part 3. 
#### Compute
 $$\int_0^1 e^{-\frac{x^2}{2}} dx$$
#### using quadrature formulas $(*)$.

In [10]:
def f(x):
    return np.exp(-x**2/2)
    
def d_f(x):
    return (-x)*np.exp(-x**2/2)

In [11]:
alpha_1 = 2/3
alpha_2 = 1/3
alpha_3 = 1/6
approximation = alpha_1*f(0) + alpha_2 * f(1) + alpha_3*d_f(0)
print("The result of the integral is", approximation)

The result of the integral is 0.8688435532375445


### Part 4. 
#### Error Estimate for quadrature formula

---

By Part 2, we know that
$$E(f)= I(f) - \hat{I}(f) =-{1\over 72}f^{(3)}(\xi)\;\;\;for\;some\;\;\xi\in[0,1]$$
$$\implies |E(f)| = \left| \int_a^b f(x)\,dx - \hat{I}(f) \right| \leq \frac{1}{72}\,K_3$$
where $|f^{(3)} (x)|\,\leq\,K_3$ for all $x \in [0,1],$  

and $$f(x)=e^{-{x^2\over 2}}\;\;\implies f^{(3)}(x)=(3x-x^3)e^{-{x^2\over 2}}\implies |f^{(3)} (x)|\,\leq\,1.5 \;\;\;\forall x \in [0,1]$$
So $$|I(f)-\hat{I}(f)|\leq \frac{1}{72}\cdot 1.5 \approx 0.0208  $$  

$$\implies \hat{I}(f)-0.0208 \leq I(f) \leq \hat{I}(f)+0.0208$$  

Plug in $\hat{I}(f)=0.8688435532375445$ from quadrature formula, we get  

$$0.8480435532375444 \leq I(f) \leq 0.8896435532375445$$